Experimental support for implicitly opening existential arguments.

DougGregor opened the pull request:
When calling a generic function with an argument of existential type,
implicitly "open" the existential type into a concrete archetype, which
can then be bound to the generic type. This extends the implicit
opening that is performed when accessing a member of an existential
type from the "self" parameter to all parameters. For example:

    func unsafeFirst<C: Collection>(_ c: C) -> C.Element { c.first! }

    func g(c: any Collection) {
      unsafeFirst(c)   // currently an error
                       // with this change, succeeds and produces an 'Any'
    }

This avoids many common sources of errors of the form

    protocol 'P' as a type cannot conform to the protocol itself

which come from calling generic functions with an existential, and
allows another way "out" if one has an existention and needs to treat
it generically.

This feature is behind a frontend flag
`-enable-experimental-opened-existential-types`.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Erase opaque types involving opened existential types to their upper bound.

AnthonyLatsis commented:
Awesome! And `for` loops could support this too!

This feature follows analogous rules to SE-309, in the sense that it can only be done if the targeted generic parameter does not appear in non-covariant position anywhere else in the type signature, right?

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
> Awesome! And `for` loops could support this too!

We certainly could!

> This feature follows analogous rules to SE-309, in the sense that it can only be done if the targeted generic parameter does not appear in non-covariant position anywhere else in the type signature, right?

I believe so, but I haven't thought through this completely yet. `findExistentialSelfReferences` likely needs to be generalized to any generic parameter.

DougGregor committed:
Opaque types are only invariant when they involve same-type constraints.

An opaque type is only invariant with respect to the existential Self
when the constraints on the opaque type involve Self. Such constraints
are not expressible in the type-erased value, so treat them as
invariant. This loosens the restriction on using members of protocol
type that return an opaque type, such that (e.g.) the following is
still well-formed:

    protocol P { }
    protocol Q { }

    extension P {
      func getQ() -> some Q { ... }
    }

    func test(p: any P) {
      let q = p.getQ() // formerly an error, now returns an "any Q"
    }

However, this does not permit uses of members such as:

    extension P {
      func getCollection() -> some Collection<Self> { ... } // error
    }

because the type system cannot express the corresponding existential
type `any Collection<Self>`.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Fix test availability for opaque result types

AnthonyLatsis commented on lib/AST/Decl.cpp:
```diff
@@ -3885,10 +3885,29 @@ findExistentialSelfReferences(CanGenericSignature existentialSig, Type type,
-  // reference to 'Self'.
-  if (type->is<OpaqueTypeArchetypeType>())
-    return SelfReferenceInfo::forSelfRef(TypePosition::Invariant);
+  // If the signature of an opaque result type has a same-type constraint
+  // that refereces Self, it's invariant.
+  if (auto opaque = type->getAs<OpaqueTypeArchetypeType>()) {
```
Nice catch! I think we can ignore opaque types altogether: it looks like the code you added to `typeEraseCovariant` would also erase stuff like `some Collection<Self>` to `any Collection`.

DougGregor replied:
Did this in https://github.com/apple/swift/pull/41183/commits/98ecb9b2029568ca9d7c7db68af34c027a932891

DougGregor replied:
It doesn't, because there's a same-type constraint between the opaque result type's `Element` type and `Self` that involves `Self`, so we reject attempts to open the existential when calling a function that returns `some Collection<Self>`.

AnthonyLatsis replied:
> so we reject attempts to open the existential when calling a function that returns `some Collection<Self>`

Type-erasing the resulting opaque archetype to `any Collection` sounds safe as far as I'm concerned. Am I missing the point, or is it something we can do in a follow-up PR?

DougGregor replied:
My concern is that we would erase to `any Collection` today, but then when we have existentials like `any Collection<String>`, we'll break code by changing the result of type inference here.

DougGregor committed:
Prevent opening existentials when the corresponding parameter is variant

Ensure that we only open existentials in an argument when the corresponding
parameter's type is a generic parameter that is only used in covariant
positions, because either invariant or contravariant uses mean that we
won't be able to type-erase other uses of that parameter. This mirrors
the requirement placed when opening existentials as a member of protocol
type.

DougGregor commented:
> `findExistentialSelfReferences` likely needs to be generalized to any generic parameter.

This is done in the 6th commit (https://github.com/apple/swift/pull/41183/commits/98ecb9b2029568ca9d7c7db68af34c027a932891), so we diagnose any contravariant or invariant uses of the generic parameter that's being bound to the opened type.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Support opening of arguments of existential metatype type.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Don't open existential arguments to `type(of:)`.

DougGregor committed:
Ensure that we only open existential arguments to functions/subscripts.

DougGregor committed:
Don't open existentials when calling C++ function templates

C++ function templates require specialization, which does not work with
opened existentials. Disable opening for them.

DougGregor committed:
SwiftOnoneSupport creates different specializations with opened existentials

Implicitly disable implicit opening of existentials for SwiftOnoneSupport
because it generates different specializations.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor committed:
Opened existential types aren't always `ExistentialType`.

As with many other places in the frontend, grab the constraint type
when from an existential type when we have one.

DougGregor committed:
Only open existentials for generic arguments on the called declaration

DougGregor committed:
Cannot open existentials passed to variadic parameters.

DougGregor committed:
Allow opening existentials for inout parameters.

DougGregor committed:
Allow existential opening for parameters of optional type.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@DougGregor Are we ready to merge this? If so, I am willing to let you merge this before #41201 so that you don't run into a conflict.

DougGregor committed:
Experimental support for implicitly opening existential arguments.

When calling a generic function with an argument of existential type,
implicitly "open" the existential type into a concrete archetype, which
can then be bound to the generic type. This extends the implicit
opening that is performed when accessing a member of an existential
type from the "self" parameter to all parameters. For example:

    func unsafeFirst<C: Collection>(_ c: C) -> C.Element { c.first! }

    func g(c: any Collection) {
      unsafeFirst(c)   // currently an error
                       // with this change, succeeds and produces an 'Any'
    }

This avoids many common sources of errors of the form

    protocol 'P' as a type cannot conform to the protocol itself

which come from calling generic functions with an existential, and
allows another way "out" if one has an existention and needs to treat
it generically.

This feature is behind a frontend flag
`-enable-experimental-opened-existential-types`.

DougGregor committed:
Add "disable" counterpart to `-enable-experimental-opened-existential-types`

DougGregor committed:
Revert "Allow existential opening for parameters of optional type."

This reverts commit 6e7fff7e65283ae9b25116fa6b75ba92fd2f2a58. The
asymmetry between opening for optional parameters and optional
arguments is surprising enough that we're currently opting not to
allow them.

DougGregor commented:
@swift-ci please smoke test
